pkg/helpers/tokencmd/negotiator_gssapi.go

```diff
@@ -4,195 +4,91 @@
-	loadOnce sync.Once
-
-	// service name of the server we are negotiating against
-	name *gssapi.Name
-	// security context holding the state of the negotiation
-	ctx *gssapi.CtxId
-	// flags indicating what options we want for the negotiation
-	// TODO: surface option for mutual authentication, e.g. gssapi.GSS_C_MUTUAL_FLAG
-	flags uint32
-
 	// principalName contains the name of the principal desired by the user, if specified.
 	principalName string
-	// cred holds the credentials obtained for the specified principalName.
-	// if no principalName is specified, GSS_C_NO_CREDENTIAL is used.
-	cred *gssapi.CredId
 
 	// track whether the last response from InitSecContext was GSS_S_COMPLETE
 	complete bool
+
+	client *krb5client.Client
 }
 
 func NewGSSAPINegotiator(principalName string) Negotiator {
 	return &gssapiNegotiator{principalName: principalName}
 }
 
 func (g *gssapiNegotiator) Load() error {
-	_, err := g.loadLib()
-	return err
+	return nil
 }
 
 func (g *gssapiNegotiator) InitSecContext(requestURL string, challengeToken []byte) (tokenToSend []byte, err error) {
-	lib, err := g.loadLib()
-	if err != nil {
-		return nil, err
-	}
-
-	// Initialize our context if we haven't already
-	if g.ctx == nil {
-
-		if len(g.principalName) > 0 {
-			// Get credentials for a specific principal
-			klog.V(5).Infof("acquiring credentials for principal name %s", g.principalName)
-			credBuffer, err := lib.MakeBufferString(g.principalName)
-			if err != nil {
-				return nil, convertGSSAPIError(err)
-			}
-			defer credBuffer.Release()
-
-			credName, err := credBuffer.Name(lib.GSS_KRB5_NT_PRINCIPAL_NAME)
-			if err != nil {
-				return nil, convertGSSAPIError(err)
-			}
-			defer credName.Release()
-
-			cred, _, _, err := lib.AcquireCred(credName, time.Duration(0), lib.GSS_C_NO_OID_SET, gssapi.GSS_C_INITIATE)
-			if err != nil {
-				klog.V(5).Infof("AcquireCred returned error: %v", err)
-				return nil, convertGSSAPIError(err)
-			}
-			g.cred = cred
-		} else {
-			// otherwise, express no opinion about the credentials and let gssapi decide
-			g.cred = lib.GSS_C_NO_CREDENTIAL
+	if g.client == nil {
+		krb5ConfLocation := krb5Config
+		if os.Getenv(krb5ConfigEnvKey) != "" {
+			krb5ConfLocation = os.Getenv(krb5ConfigEnvKey)
 		}
 
-		serviceName, err := getServiceName('@', requestURL)
+		config, err := krb5conf.Load(krb5ConfLocation)
 		if err != nil {
 			return nil, err
 		}
 
-		klog.V(5).Infof("importing service name %s", serviceName)
-		nameBuf, err := lib.MakeBufferString(serviceName)
+		krb5KeytabLocation := krb5Keytab
+		if os.Getenv(krb5KeytabEnvKey) != "" {
+			krb5KeytabLocation = os.Getenv(krb5KeytabEnvKey)
+		}
+		keytab, err := krb5kt.Load(krb5KeytabLocation)
 		if err != nil {
-			return nil, convertGSSAPIError(err)
+			return nil, err
 		}
-		defer nameBuf.Release()
 
-		name, err := nameBuf.Name(lib.GSS_C_NT_HOSTBASED_SERVICE)
+		g.client = krb5client.NewWithKeytab(g.principalName, config.LibDefaults.DefaultRealm, keytab, config)
```

Comment:
@LorbusChris as far as I can see, realm is required. It seems to be safe to me to use `LibDefaults.DefaultRealm` in this case, what do you think?

Reply:
agreed :+1: